nativelib/src/main/resources/gc/commix/Marker.c

```diff
@@ -0,0 +1,367 @@
+                        GreyPacket_Push(slice, object);
+                        Marker_giveFullPacket(heap, stats, slice);
+                    }
+                }
+            }
+            // non-object arrays do not contain pointers
+        } else {
+            int64_t *ptr_map = object->rtti->refMapStruct;
+            int i = 0;
+            while (ptr_map[i] != LAST_FIELD_OFFSET) {
+                word_t *field = object->fields[ptr_map[i]];
+                if (Heap_IsWordInHeap(heap, field)) {
+                    ObjectMeta *fieldMeta = Bytemap_Get(bytemap, field);
+                    if (ObjectMeta_IsAllocated(fieldMeta)) {
+                        Marker_markObject(heap, stats, outHolder, bytemap, (Object *)field,
+                                          fieldMeta);
+                    }
+                }
+                ++i;
+            }
+        }
+    }
+}
+
+void Marker_markRangePacket(Heap *heap, Stats *stats, GreyPacket* in, GreyPacket **outHolder) {
+    Bytemap *bytemap = heap->bytemap;
+    if (*outHolder == NULL) {
+        GreyPacket *fresh = Marker_takeEmptyPacket(heap, stats);
+        assert(fresh != NULL);
+        *outHolder = fresh;
+    }
+    word_t **fields = (word_t **) in->items[0];
+    Marker_markRange(heap, stats, in, outHolder, bytemap, fields, ARRAY_SPLIT_BATCH);
+    in->type = grey_packet_reflist;
+    in->size = 0;
+}
+
+static inline void Marker_markBatch(Heap *heap, Stats *stats, GreyPacket* in, GreyPacket **outHolder) {
+#ifdef ENABLE_GC_STATS_BATCHES
+    uint64_t start_ns, end_ns;
+    if (stats != NULL) {
+        start_ns = scalanative_nano_time();
+    }
+#endif
+    switch (in->type) {
+        case grey_packet_reflist:
+            Marker_markPacket(heap, stats, in, outHolder);
+            break;
+        case grey_packet_refrange:
+            Marker_markRangePacket(heap, stats, in, outHolder);
+            break;
+    }
+#ifdef ENABLE_GC_STATS_BATCHES
+    if (stats != NULL) {
+        end_ns = scalanative_nano_time();
+        Stats_RecordEvent(stats, event_mark_batch, start_ns, end_ns);
+    }
+#endif
+}
+
+void Marker_Mark(Heap *heap, Stats *stats) {
+    GreyPacket* in = Marker_takeFullPacket(heap, stats);
+    GreyPacket *out = NULL;
+    while (in != NULL) {
+        Marker_markBatch(heap, stats, in, &out);
+
+        assert(out != NULL);
+        assert(GreyPacket_IsEmpty(in));
+        GreyPacket *next = Marker_takeFullPacket(heap, stats);
+        if (next != NULL) {
+            Marker_giveEmptyPacket(heap, stats, in);
+        } else {
+            if (!GreyPacket_IsEmpty(out)) {
+                // use the out packet as source
+                next = out;
+                out = in;
+            } else {
+                // next == NULL, exits
+                Marker_giveEmptyPacket(heap, stats, in);
+                Marker_giveEmptyPacket(heap, stats, out);
+            }
+        }
+        in = next;
+    }
+}
+
+void Marker_MarkAndScale(Heap *heap, Stats *stats) {
```

Comment:
1. Needs a comment with motivation for not using all threads at once. 
2. Some of thread management boilerplate probably belongs to non-marker module (i.e. gcthread module).